File: src/compiler/parser/UserException.java
package compiler.parser;

import compiler.EntryPoint;
import compiler.lexer.token.Token;
import compiler.utils.TextCursor;

public class UserException extends RuntimeException {
  public UserException(String message, Token token) {
    super(message + "\n" + renderLineNumber(token, EntryPoint.inputName));
  }

  public UserException(String message, TextCursor token) {
    super(message + "\n" + renderLineNumber(token, EntryPoint.inputName));
  }

  private static String renderLineNumber(Token token, String inputName) {
    return "\tat " + inputName + "(" + inputName + ":" + token.getLineNumber() + ")";
  }

  private static String renderLineNumber(TextCursor cursor, String inputName) {
    return "\tat " + inputName + "(" + inputName + ":" + cursor.getCursorLineNumber() + ")";
  }
}

File: src/compiler/parser/UserException.kt
package compiler.parser

import compiler.EntryPoint
import compiler.parser.Language.Token
import compiler.utils.TextCursor

open class UserException : RuntimeException {
    constructor(message: String, token: Token) : super(message + "\n" + renderLineNumber(token, EntryPoint.inputName))
    constructor(message: String, cursor: TextCursor) : super(message + "\n" + renderLineNumber(cursor, EntryPoint.inputName))

    companion object {
        private fun renderLineNumber(token: Token, inputName: String?): String {
            return "\tat " + inputName + "(" + inputName + ":" + token.lineInfo.number + ")"
        }

        private fun renderLineNumber(cursor: TextCursor, inputName: String?): String {
            return "\tat " + inputName + "(" + inputName + ":" + cursor.getCursorLineNumber() + ")"
        }
    }
}
